dask_match/repartition.py

```diff
@@ -1,22 +1,113 @@
+        if self.n is not None:
+            if self.n < self.frame.npartitions:
+                return RepartitionToFewer(self.frame, self.n)
+            else:
+                original_divisions = divisions = pd.Series(
+                    self.frame.divisions
+                ).drop_duplicates()
+                if self.frame.known_divisions and (
+                    is_datetime64_any_dtype(divisions.dtype)
+                    or is_numeric_dtype(divisions.dtype)
+                ):
+                    npartitions = self.n
+                    df = self.frame
+                    if is_datetime64_any_dtype(divisions.dtype):
+                        divisions = divisions.values.astype("float64")
+
+                    if is_series_like(divisions):
+                        divisions = divisions.values
+
+                    n = len(divisions)
+                    divisions = np.interp(
+                        x=np.linspace(0, n, npartitions + 1),
+                        xp=np.linspace(0, n, n),
+                        fp=divisions,
+                    )
+                    if is_datetime64_any_dtype(original_divisions.dtype):
+                        divisions = methods.tolist(
+                            pd.Series(divisions).astype(original_divisions.dtype)
+                        )
+                    elif np.issubdtype(original_divisions.dtype, np.integer):
+                        divisions = divisions.astype(original_divisions.dtype)
+
+                    if isinstance(divisions, np.ndarray):
+                        divisions = divisions.tolist()
+
+                    divisions = list(divisions)
+                    divisions[0] = df.divisions[0]
+                    divisions[-1] = df.divisions[-1]
+
+                    # Ensure the computed divisions are unique
+                    divisions = list(unique(divisions[:-1])) + [divisions[-1]]
+                    return RepartitionDivisions(df, divisions, self.force)
+                else:
+                    return RepartitionToMore(self.frame, self.n)
+        elif self.new_divisions:
+            return RepartitionDivisions(self.frame, self.new_divisions, self.force)
+        else:
+            raise NotImplementedError()
+
+    @classmethod
+    def _replacement_rules(cls):
+        # Move filter or column projection before repartitioning
+        frame, other = map(Wildcard.dot, ["frame", "other"])
+
+        def reorder(frame, other, op=Filter):
+            return type(frame)(op(frame.operands[0], other), *frame.operands[1:])
+
+        supported = CustomConstraint(
+            lambda frame, other: (
+                # filter-pushdown gets messy if other is an Expr
+                isinstance(frame, cls)
+                and not isinstance(other, Expr)
+            )
+        )
+        for op in [Filter, Projection]:
+            pattern = Pattern(op(frame, other), supported)
+            yield ReplacementRule(pattern, functools.partial(reorder, op=op))
```

Comment:
Will need to rewrite this with `simplify_up` - So, I'd say we should treat #50 as a dependency for this PR.

Reply:
Done.